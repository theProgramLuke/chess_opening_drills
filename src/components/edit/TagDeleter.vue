<template lang="pug">
  v-dialog(v-model="showDialog" max-width="750px")
    template(v-slot:activator="{on: dialog, attrs}")
      v-tooltip(bottom)
        template(v-slot:activator="{on: tooltip }")
          v-btn(v-bind="attrs", v-on="{ ...tooltip, ...dialog }", icon, color="error", :disabled="disabled")
            v-icon mdi-delete

        span Delete tag

    v-card.pa-4
      v-card-title Delete tag?
      v-card-subtitle The tag "{{ tag.name }}" and all of its child tags will be permanently deleted.
      v-form(@submit.prevent="onDelete")
        v-card-actions
          v-btn.ma-2(type="submit", color="error")
            | Delete
          v-btn.ma-2(@click="showDialog=false", color="secondary")
            | Cancel
</template>

<script lang="ts" src="./TagDeleterViewModel.ts" />
