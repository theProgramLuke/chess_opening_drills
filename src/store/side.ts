export enum Side {
  White,
  Black
}
