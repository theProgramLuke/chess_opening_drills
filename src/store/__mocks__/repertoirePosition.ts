import _ from "lodash";

import { Move } from "../move";
import { Side } from "../side";
import { TrainingEvent } from "../TrainingEvent";
import { TrainingMode } from "../trainingMode";
import { Turn } from "../turn";
import { FEN } from "chessground/types";
import { SavedRepertoirePosition } from "../repertoirePosition";

export class RepertoirePosition {
  fen: string;
  comment: string;
  parents: RepertoirePosition[];
  children: Move[];
  myTurn: boolean;
  forSide: Side;
  trainingHistory: TrainingEvent[];
  nextRepetitionTimestamp?: number | undefined;
  previousIntervalDays: number;
  intervalIndex: number;
  easinessFactor: number;
  SideToMove: () => Side;
  AddChild: (move: Move) => void;
  Unlink: () => void;
  IncludeForTrainingMode: (
    mode: TrainingMode,
    difficultyModeLimit?: number
  ) => boolean;
  AddTrainingEvent: (event: TrainingEvent) => void;
  VisitChildren: (visit: (child: RepertoirePosition) => void) => void;
  IsChildPosition: (maybeChild: RepertoirePosition) => boolean;
  RootPaths: () => Move[][];
  GetTurnLists: () => Turn[][];
  AsPgn: () => string;
  AsPgnMoveText: (turnCount?: number) => string;
  AsSaved: (childParentSource: RepertoirePosition[]) => SavedRepertoirePosition;

  constructor(fen: FEN, comment: string, forSide: Side) {
    this.fen = fen;
    this.comment = comment;
    this.forSide = forSide;
    this.parents = [];
    this.children = [];
    this.myTurn = false;
    this.forSide = Side.White;
    this.trainingHistory = [];
    this.nextRepetitionTimestamp = 0;
    this.previousIntervalDays = 0;
    this.intervalIndex = 0;
    this.easinessFactor = 0;
    this.SideToMove = jest.fn();
    this.AddChild = jest.fn();
    this.Unlink = jest.fn();
    this.IncludeForTrainingMode = jest.fn();
    this.AddTrainingEvent = jest.fn();
    this.VisitChildren = jest.fn();
    this.IsChildPosition = jest.fn();
    this.RootPaths = jest.fn();
    this.GetTurnLists = jest.fn();
    this.AsPgn = jest.fn();
    this.AsPgnMoveText = jest.fn();
    this.AsSaved = jest.fn();
  }
}
