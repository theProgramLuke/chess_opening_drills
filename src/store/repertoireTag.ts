import { Side } from "./side";

export class RepertoireTag {
  id: number;
  forSide: Side;
  name: string;
  position: string;
  children: Array<RepertoireTag>;

  constructor(
    id: number,
    forSide: Side,
    name: string,
    position: string,
    children: Array<RepertoireTag>
  ) {
    this.id = id;
    this.forSide = forSide;
    this.name = name;
    this.position = position;
    this.children = children;
  }
}

function _findTag(tag: RepertoireTag, id: number): RepertoireTag | null {
  if (tag["id"] === id) {
    return tag;
  }

  if (tag["children"].length === 0) {
    return null;
  }

  const mapped = tag["children"].map((child: RepertoireTag) =>
    _findTag(child, id)
  );
  const filtered = mapped.filter((child: RepertoireTag | null) => {
    return child != null;
  });

  return filtered[0];
}

export function FindRepertoireTag(
  tags: Array<RepertoireTag>,
  id: number
): RepertoireTag | null {
  const mapped = tags.map((child: RepertoireTag) => _findTag(child, id));
  const filtered = mapped.filter((child: RepertoireTag | null) => {
    return child != null;
  });

  // assert(filtered.length === 1, "Should only find one matching tag...");

  return filtered[0];
}
