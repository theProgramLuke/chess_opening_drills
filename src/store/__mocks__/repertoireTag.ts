import { FEN } from "chessground/types";

import * as Real from "@/store/repertoireTag";
import { Side } from "@/store/side";
import { RepertoirePosition } from "@/store/repertoirePosition";

export class RepertoireTag {
  forSide: Side;
  name: string;
  position: RepertoirePosition;
  fen: FEN;
  children: RepertoireTag[];
  id: string;

  constructor(
    forSide: Side,
    name: string,
    position: RepertoirePosition,
    fen: FEN,
    children: RepertoireTag[],
    id?: string
  ) {
    this.id = id || "";
    this.forSide = forSide;
    this.name = name;
    this.position = position;
    this.fen = fen;
    this.children = children;
    this.AddChild = jest.fn();
    this.RemoveChild = jest.fn();
    this.AsSaved = jest.fn();
  }

  AddChild: (tag: RepertoireTag) => void;

  RemoveChild: (tag: RepertoireTag) => void;

  AsSaved: (positionSource: RepertoirePosition[]) => Real.SavedRepertoireTag;

  static FromSaved: (
    saved: Real.SavedRepertoireTag,
    positionSource: RepertoirePosition[]
  ) => RepertoireTag = jest.fn();
}

export const GetTrainingPositions = jest.fn();

export const GetTrainingMoveLists = jest.fn();
