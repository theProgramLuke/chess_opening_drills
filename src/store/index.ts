import Vue from "vue";
import Vuex from "vuex";

import { RepertoirePosition } from "@/store/repertoirePosition";
import { Move } from "./move";
import { GetPersistantStorage } from "./storage";
import { Repertoire } from "./repertoire";
import { RepertoireTag } from "./repertoireTag";
import { Side } from "./side";
import { TrainingEvent } from "./TrainingEvent";
import { PgnGame } from "./pgnParser";

Vue.use(Vuex);

const storage = GetPersistantStorage();

export default new Vuex.Store({
  state: {
    darkMode: storage.get("darkMode"),
    primary: storage.get("primary"),
    secondary: storage.get("secondary"),
    accent: storage.get("accent"),
    error: storage.get("error"),
    warning: storage.get("warning"),
    info: storage.get("info"),
    success: storage.get("success"),
    boardTheme: storage.get("boardTheme"),
    pieceTheme: storage.get("pieceTheme"),
    whiteRepertoire: Repertoire.FromSaved(storage.get("whiteRepertoire")),
    blackRepertoire: Repertoire.FromSaved(storage.get("blackRepertoire"))
  },

  mutations: {
    setDarkMode: (state, darkMode): void => {
      state.darkMode = darkMode;
      storage.set("darkMode", darkMode);
    },

    setBoardTheme: (state, boardTheme): void => {
      state.boardTheme = boardTheme;
      storage.set("boardTheme", boardTheme);
    },

    setPieceTheme: (state, pieceTheme): void => {
      state.pieceTheme = pieceTheme;
      storage.set("pieceTheme", pieceTheme);
    },

    setColor: function(
      state,
      payload: {
        colorToSet:
          | "primary"
          | "secondary"
          | "accent"
          | "error"
          | "warning"
          | "info"
          | "success";
        color: string;
      }
    ): void {
      state[payload.colorToSet] = payload.color;
      storage.set(payload.colorToSet, payload.color);
    },

    addRepertoirePosition: (
      state,
      payload: { parent: RepertoirePosition; newMove: Move }
    ): void => {
      if (payload.parent && payload.newMove) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddMove(payload.parent, payload.newMove);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addRepertoireTag: (
      state,
      payload: { parent: RepertoireTag; tag: RepertoireTag }
    ): void => {
      if (payload.parent && payload.tag) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.parent.AddChild(payload.tag);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addTrainingEvent: (
      state,
      payload: { position: RepertoirePosition; event: TrainingEvent }
    ): void => {
      if (payload.position && payload.event) {
        const repertoire =
          payload.position.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.position.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.position.AddTrainingEvent(payload.event);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    removeRepertoireTag: (state, tag: RepertoireTag): void => {
      const repertoire =
        tag.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        tag.forSide === Side.White ? "whiteRepertoire" : "blackRepertoire";

      repertoire.RemoveRepertoireTag(tag);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    removeRepertoireMove: (state, move: Move): void => {
      const repertoire =
        move.position.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        move.position.forSide === Side.White
          ? "whiteRepertoire"
          : "blackRepertoire";

      repertoire.RemoveMove(move);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    addPositionsFromGame(
      state,
      payload: { forSide: Side; game: PgnGame }
    ): void {
      if (payload.game) {
        const repertoire =
          payload.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddFromGame(payload.game);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    clearStorage(): void {
      storage.clear();
    }
  },

  actions: {},

  modules: {},

  strict: process.env.NODE_ENV !== "production"
});
