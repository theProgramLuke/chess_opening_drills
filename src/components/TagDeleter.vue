<template lang="pug">
  v-dialog(v-model="showDialog" max-width="500px")
    template(v-slot:activator="{on, attrs}")
      v-btn(v-bind="attrs" v-on="on" icon, color="error", :disabled="disabled")
        v-icon mdi-delete

    v-card.pa-4
      v-card-title Delete?
      v-card-subtitle The tag "{{ tag.name }}" and all of its child tags will be permanently deleted.
      v-btn.ma-2(@click="onDelete(), showDialog=false", color="error") Delete
      v-btn.ma-2(@click="showDialog=false", color="secondary", text, outlined) Cancel
</template>

<script lang="ts">
import Vue from "vue";
import { mapState, mapMutations } from "vuex";
import { RepertoireTag } from "@/store/repertoireTag";

export default Vue.extend({
  data: () => ({
    showDialog: false
  }),
  props: {
    tag: {
      type: RepertoireTag,
      required: true
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    onDelete() {
      this.$emit("onDelete", this.tag);
    }
  }
});
</script>
