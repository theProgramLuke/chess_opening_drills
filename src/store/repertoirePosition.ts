import _ from "lodash";
import { FEN } from "chessground/types";
import { Chess } from "chess.js";

import { Move, SavedMove } from "@/store/move";
import { Side } from "@/store/side";
import { Turn } from "@/store/turn";

export class SavedRepertoirePosition {
  fen: FEN;
  comment: string;
  parentIds: number[];
  children: SavedMove[];
  myTurn: boolean;
  forSide: Side;

  constructor(
    fen: FEN,
    comment: string,
    parentIds: number[],
    children: SavedMove[],
    myTurn: boolean,
    forSide: Side
  ) {
    this.fen = fen;
    this.comment = comment;
    this.parentIds = parentIds;
    this.children = children;
    this.myTurn = myTurn;
    this.forSide = forSide;
  }
}

export class RepertoirePosition {
  fen: FEN;
  comment: string;
  parents: RepertoirePosition[];
  children: Move[];
  myTurn: boolean;
  forSide: Side;

  constructor(fen: FEN, myTurn: boolean, comment: string, forSide: Side) {
    this.fen = fen;
    this.comment = comment;
    this.myTurn = myTurn;
    this.forSide = forSide;
    this.parents = [];
    this.children = [];
  }

  SideToMove(): Side {
    return Chess(this.fen).turn() === "w" ? Side.White : Side.Black;
  }

  AddChild(move: Move): void {
    const samePosition = this.fen == move.position.fen;
    const alreadyChild = !_.isEmpty(
      _.find(
        this.children,
        (child: Move) => move.position.fen === child.position.fen
      )
    );
    if (!samePosition && !alreadyChild) {
      this.children.push(move);
      move.position.parents.push(this);
    }
  }

  private RootPaths(): Move[][] {
    const collector: Move[][] = [];

    if (!_.isEmpty(this.parents)) {
      this.RootPathsRecursive([], collector);
    }

    return collector;
  }

  private RootPathsRecursive(path: Move[], collector: Move[][]): void {
    if (_.isEmpty(this.parents)) {
      collector.push(path);
    } else {
      _.forEach(this.parents, parent => {
        const branch = _.clone(path);
        const parentMove = parent.getChildMove(this);

        if (parentMove) {
          branch.unshift(parentMove);
        }

        parent.RootPathsRecursive(branch, collector);
      });
    }
  }

  private getChildMove(other: RepertoirePosition): Move | undefined {
    return _.find(this.children, child => child.position === other);
  }

  GetTurnLists(): Array<Array<Turn>> {
    const moveLists: Array<Array<Turn>> = [];

    const paths = this.RootPaths();

    _.forEach(paths, (path: Array<Move>) => {
      const turns: Array<Turn> = [];

      _.forEach(_.range(0, path.length, 2), (i: number) => {
        const whiteMove = path[i];

        if (path.length === i) {
          turns.push(new Turn(whiteMove));
        } else {
          const blackMove = path[i + 1];
          turns.push(new Turn(whiteMove, blackMove));
        }
      });

      moveLists.push(turns);
    });

    return moveLists;
  }

  AsPgnMoveText(turnCount = 1): string {
    if (_.isEmpty(this.children)) {
      return "";
    }

    const side = this.SideToMove();
    let pgnMoveText = "";
    let childTurnCount = turnCount;

    if (side === Side.White) {
      // White made this move.
      pgnMoveText += turnCount + ". ";
      childTurnCount++;
    }

    const mainLine = _.first(this.children);
    if (mainLine) {
      pgnMoveText += mainLine.san;
    }

    const variations = _.tail(this.children);

    _.forEach(variations, child => {
      pgnMoveText += " ( "; // start variation
      pgnMoveText += turnCount;
      pgnMoveText += side === Side.White ? ". " : "... ";
      pgnMoveText += child.san;
      pgnMoveText += " ";
      pgnMoveText += child.position.AsPgnMoveText(childTurnCount);
      pgnMoveText += " )"; // end variation
    });

    if (mainLine) {
      pgnMoveText += " ";
      pgnMoveText += mainLine.position.AsPgnMoveText(childTurnCount);
    }

    return _.trim(pgnMoveText);
  }

  AsSaved(childParentSource: RepertoirePosition[]): SavedRepertoirePosition {
    const savedChildren = _.map(
      this.children,
      child =>
        new SavedMove(child.san, _.indexOf(childParentSource, child.position))
    );

    const parentIds = _.map(this.parents, parent =>
      _.indexOf(childParentSource, parent)
    );

    return new SavedRepertoirePosition(
      this.fen,
      this.comment,
      parentIds,
      savedChildren,
      this.myTurn,
      this.forSide
    );
  }

  static FromSaved(saved: SavedRepertoirePosition): RepertoirePosition {
    return new RepertoirePosition(
      saved.fen,
      saved.myTurn,
      saved.comment,
      saved.forSide
    );
  }
}
