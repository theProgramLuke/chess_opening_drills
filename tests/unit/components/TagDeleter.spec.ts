import { shallowMount } from "@vue/test-utils";
import vuetify from "vuetify";
import Vue from "vue";

import TagDeleter from "@/components/TagDeleter.vue";
import { RepertoireTag } from "@/store/repertoireTag";
import { Side } from "@/store/side";
import { RepertoirePosition } from "@/store/repertoirePosition";

beforeEach(() => {
  Vue.use(vuetify);
});

describe("TagDeleter", () => {
  it("renders correctly", () => {
    const component = shallowMount(TagDeleter, {
      propsData: {
        tag: new RepertoireTag(
          Side.White,
          "tag name",
          new RepertoirePosition("", "", Side.White),
          "",
          []
        )
      }
    });

    expect(component.element).toMatchSnapshot();
  });
});
