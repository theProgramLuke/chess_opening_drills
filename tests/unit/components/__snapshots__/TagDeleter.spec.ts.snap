// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagDeleter renders correctly 1`] = `
<v-dialog-stub
  closedelay="0"
  contentclass=""
  maxwidth="750px"
  opendelay="0"
  origin="center center"
  retainfocus="true"
  transition="dialog-transition"
  width="auto"
>
  <v-card-stub
    class="pa-4"
    loaderheight="4"
    tag="div"
  >
    <v-card-title-stub>
      Delete tag?
    </v-card-title-stub>
    <v-card-subtitle-stub>
      The tag "tag name" and all of its child tags will be permanently deleted.
    </v-card-subtitle-stub>
    <v-form-stub>
      <v-card-actions-stub>
        <v-btn-stub
          activeclass=""
          class="ma-2"
          color="error"
          tag="button"
          type="submit"
        >
          Delete
        </v-btn-stub>
        <v-btn-stub
          activeclass=""
          class="ma-2"
          color="secondary"
          tag="button"
          type="button"
        >
          Cancel
        </v-btn-stub>
      </v-card-actions-stub>
    </v-form-stub>
  </v-card-stub>
</v-dialog-stub>
`;
